test/blackbox-tests/test-cases/pkg/installed-binary.t

```diff
@@ -0,0 +1,39 @@
+Test that installed binaries are visible in dependent packages
+
+  $ mkdir dune.lock
+  $ cat >dune.lock/test <<EOF
+  > (build
+  >  (system "\| echo "#!/bin/sh\necho from test package" > foo;
+  >          "\| chmod +x foo;
+  >          "\| touch libxxx lib_rootxxx;
+  >          "\| cat >test.install <<EOF
+  >          "\| bin: [ "foo" ]
+  >          "\| lib: [ "libxxx" ]
+  >          "\| lib_root: [ "lib_rootxxx" ]
+  >          "\| share_root: [ "lib_rootxxx" ]
+  >          "\| EOF
+  >  ))
+  > EOF
+
+  $ cat >dune.lock/usetest <<EOF
```

Comment:
Is the idea that the lock file is actually a directory containing a file for each package that describes where to get the source and how to build that package?

Reply:
That's exactly it. A directory was chosen over a single file to make it easier to diff lock files.

Reply:
We discussed this at the dune-dev meeting and I think it would be helpful to discus a bit more the tradeoff between multiple files and a single file. It seems like a single file sorted by package name would also be quite easy to diff and would have additional benefits such as simplifying working with multiple lockfiles (corresponding to multiple platforms or CI experiments for example) and would also simplify some git workflows compared to one file per package.

Reply:
Sorting helps, but it's easily confused when we have many packages with similar metadata.

>  such as simplifying working with multiple lockfiles (corresponding to multiple platforms or CI experiments for example)

Could you explain this point a bit more? I don't see any difference here.

>  and would also simplify some git workflows compared to one file per package.

Which git workflows?

Reply:
> 

> Sorting helps, but it's easily confused when we have many packages with similar metadata.
> 
> > such as simplifying working with multiple lockfiles (corresponding to multiple platforms or CI experiments for example)
> 
> Could you explain this point a bit more? I don't see any difference here.

For example if we want to test with ocaml 4 and ocaml 5 in CI we might create one lockfile for each ocaml version and it seems messy to need one file for each combination of lockfile x package.

> 
> > and would also simplify some git workflows compared to one file per package.
> 
> Which git workflows?

One example would be when you add a new dependency this would create a new file and this wouldn't automatically get checked in when running `git commit -a` which is what I suspect a lot of people run.

Another issue we discussed yesterday which I forgot to mention is that if a project has many dependencies having one file per dependency could cause problems on systems with a maximum number of files per directory.

Reply:
> For example if we want to test with ocaml 4 and ocaml 5 in CI we might create one lockfile for each ocaml version and it seems messy to need one file for each combination of lockfile x package.

I still don't get it. We can create one lock "file" for each compiler version with a directory. One scheme would be `4.14.dune.lock/`, `5.0.dune.lock/` etc.

> One example would be when you add a new dependency this would create a new file and this wouldn't automatically get checked in when running git commit -a which is what I suspect a lot of people run.

Although I don't use git this way, I grant that this is an issue, but it seems like a minor inconvenience. While diffing is a big problem every time the lock file is updated.

> Another issue we discussed yesterday which I forgot to mention is that if a project has many dependencies having one file per dependency could cause problems on systems with a maximum number of files per directory.

On such systems, we're going to run into problems with cloning opam-repository first. Which has 4k entries in its packages/ directory.